text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,195 @@
+
+# Motivation
+[motivation]: #motivation
+
+Rust is very promising for security-critical applications due to its safety guarantees, but there currently are gaps in the ecosystem that prevent it. One of them is the lack of any infrastructure for security updates.
+
+Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.
+
+The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or third-party databases such as [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is `.dep-v0` across all platforms. The section name must be changed if breaking changes are made to the format.
```

Comment:
Going to re-raise a concern from earlier in the thread

> Given that this uses a completely Rust-specific (and cargo-specific) format, the section name should not use something generic like .dep-list.

(now `.dep-v0`)

Reply:
Should we separately encode schema + version + compression?

Reply:
Allowing permutations of schema+version+compression methods is an explicit non-goal. Experiments with other compression methods show that they provide little to no gain over zlib for this use case. JSON is carefully chosen to be easy to consume both programmatically and by humans in a pinch.

Allowing more format variations would also complicate parsers - they would have to support several variants of the format, which leads to more attack surface and more room for bugs.

Reply:
I've noted in the RFC that the exact name of the section is subject to bikeshedding.

Reply:
> Note that there is a hard limit of 8 bytes on the name

Am I right to think that that still would allow, e.g., `.rsdeps0` (or `.rsDeps0` if upper-case letters work)?  If a version number > 9 is needed, the "s" could be dropped.

Reply:
> Note that there is a hard limit of 8 bytes on the name

Am I right to think that that still would allow, e.g., `.rsdeps0` (or `.rsDeps0` if upper-case letters work)? If a version number > 9 is needed, `deps` could be shortened further to `dep`.

Reply:
Yes, that is permitted by the section naming constraints.